### PR TITLE
Add template types to `Connection`

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
Making `BaseConnection` a generic type, thus extensions need to specify the type of the connection.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Database/BaseConnection.php [hunk 1]
```diff
@@ -45,6 +45,10 @@
  * @property int        $transDepth
  * @property bool       $transFailure
  * @property bool       $transStatus
+ *
+ * @template TConnection of bool|object|resource
```
Comment (MGatner) [inline system/Database/BaseConnection.php:49]: You're going to need to explain the `bool` to me... Is this to support returning `false` on failure?

**Review by MGatner — COMMENTED** [review]
Where did those abstract functions go?

**Comment by paulbalandan** [comment]
> Where did those abstract functions go?

They are actually defined by `ConnectionInterface` and unnecessarily duplicated as `abstract public` in `BaseConnection`.

File: system/Database/BaseConnection.php (hunk 1 above)
Comment (paulbalandan) [reply system/Database/BaseConnection.php:49]: On second look, I think this should have been `false` not `bool`.

**Comment by paulbalandan** [comment]
I'll fix the conflicts once #6713 is merged.

**Comment by MGatner** [comment]
Thanks for explaining (and extracting).

**Review by kenjis — APPROVED** [review]

**Review by MGatner — APPROVED** [review]
Amazing cleanup of the baseline!

**Pull request merged**